### DISCUSSION TITLE
Difficulty setting non-trivial PenaltyDirichletBC with constant monomial

### DISCUSSION THREAD

**Discussion by helen-brooks** [idaholab/moose#20355] [Q&A General] [answered]
I seem to be having trouble using ADPenaltyDirichletBC and PenaltyDirichletBC with a constant monomial variable. In an essentially trivial setup (diffusion on 10x10 2D generated mesh), as soon as I set the value of the boundary condition to be anything other than identically zero I get the following error:

> linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
>                  PC failed due to FACTOR_NUMERIC_ZEROPIVOT

I've played around with both AD and non-AD versions, with various values of the penalty and the value of the BC to no avail. I have definitely managed to get something  like this working in the past, so I wonder if something has changed?

Please find my minimal example here:
```
[Mesh]
  type = GeneratedMesh
  dim = 2
  xmin = -1
  xmax = 1
  ymin = -1
  ymax = 1
  nx = 10
  ny = 10
[]

[Variables]
  [./u]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[Kernels]
  [./diff]
    type = ADDiffusion
    variable = u
  [../]
[]

[BCs]
  [./bc-top]
    type = ADPenaltyDirichletBC
    variable = u
    value = 0
    boundary = 'top'
    penalty = 1e5
  [../]
  [./bc-bot]
    type = ADPenaltyDirichletBC
    variable = u
    value = 1
    boundary = 'bottom'
    penalty = 1e5
  [../]
[]

[Executioner]
  type = Steady
  solve_type = 'NEWTON'
[]

[Outputs]
  execute_on = 'timestep_end'
  exodus = true
  csv = true
[]
```

**Comment by GiudGiud** [comment]
Hello

The preconditioner is dying, probably there is a zero pivot / saddle point?
@lindsayad I dont remember how we investigate that one?
For a problem that simple, it's likely the numerical problem is ill-posed. I think it has to do with the CONSTANT MONOMIAL variable type not having the required derivatives for the ADDiffusion kernel. This kernel does an integration by part to compute the laplacian with a first order derivative. You would need information from the neighbors to compute a first derivative, the local shape function (flat) has a 0 derivative. 

This preconditioner manages to get an answer but the answer seems wrong though. We expect a linear slope from bottom to top right? So this points to the problem I mentioned with the variable type. 
```
[Executioner]
  type = Steady
  solve_type = 'NEWTON'
  petsc_options_iname = ' -pc_type -pc_factor_shift_type '
  petsc_options_value = ' lu NONZERO '
[]
```

Could you please try with another variable type? Is there any reason you want to be using constant monomials?

Guillaume

**Reply by helen-brooks** [reply]
This wasn't for a specific physics problem - I was trying to make a minimal but non-trivial  example using constant monomials. Ironically, this was actually to exemplify to others some of the different element types available in MOOSE.

**Comment by lindsayad** [comment, marked as answer]
If you are using a discontinuous basis, then you must use discontinuous Galerkin kernels. As @GiudGiud implied, this problem is indeed very singular as setup. If you run with `-pc_type svd -pc_svd_monitor` on the command line, 80 of the 100 degrees of freedom are singular. The 20 that are not actually correspond to the degrees of freedom touched by your penalty conditions. It is the internal degrees of freedom that are singular because `_grad_test` is 0 because the basis is constant within the element, so no matter what your solution is on the interior, the residuals will be 0.

You can look in `moose/test/tests/dgkernels` for some examples of using `DGKernels` if you wish to continue using a discontinuous basis

**Reply by helen-brooks** [reply]
OK, thanks. I don't think I had appreciated the need to use DGKernels. Now that I come to think of it, I've only used const monomials in aux variables.
